texk/ps2pk/pk2bm.1
.TH PK2BM 1 "30 August 2022"
.SH NAME
pk2bm \- create a bitmap from a TeX pkfont
.SH SYNOPSIS:
pk2bm [-bh] [-W\fIwidth\fP] [-H\fIheight\fP] -c\fIchar\fP pkfont 

pk2bm [-bh] [-W\fIwidth\fP] [-H\fIheight\fP] -o\fIoctchar\fP pkfont 
.SH DESCRIPTION
This program generates a bitmap (ASCII file) which can be used
to create X11 applications. The bitmap file consists of all pixels
of the specified character (via the -c or -o option) from 
the given pkfont. The format is described in bitmap(X11).
.PP
The pkfont is a packed fontfile generated by gftopk(TeX) from a 
gffont. A gffont is the output of METAFONT a program to design fonts
in a device independent way.
.PP
With the -b flag a bitmap is generated in which all black pixels are
drawn as a `*' and all white bits as a `.'. With the -h flag a 
hexadecimal bitmap dump is generated. 
.PP
The -W and/or -H options can be used to create a bitmap of
respectivally `width' and `height' pixels. The pk-bitmap will in
this case be centered according to these new dimensions.
.PP
The output is written to the standard output.
.SH SEE ALSO
.nf
`METAFONT', Donald Knuth.
`The PKtype processor', belonging to the METAFONTware.
bitmap(X11), 
gftopk(TeX), 
pktype(TeX), 
ps2pk(1)

.SH AUTHOR
.nf
Piet Tutelaers
Maintained in TeX Live; please send bug reports or other correspondence
to tex-k@tug.org (https://lists.tug.org/tex-k).


ChangeLog
2022-03-21  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): 2020-03-21 for TL'22.

2022-01-18  Karl Berry  <karl@freefriends.org>

	* version.ac: 2022, no more /dev.

2021-10-04  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake-1.16.5.

2021-06-27  Karl Berry  <karl@freefriends.org>

	* reautoconf: pass -Wno-obsolete by default, since Autoconf 2.70
	and later obsoletes many common macros. See m4/README for more.
	Also standardize shell usage, etc., but no other functional changes.

2021-04-17  Karl Berry  <karl@tug.org>

	* version.ac: now 2022/dev since TL'21 is released.

2021-03-23  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-23 for TL'21.

2021-02-17  Karl Berry  <karl@tug.org>

	* version.ac: 2021, no more /dev.

2020-05-28  Karl Berry  <karl@freefriends.org>

	* Makefile.am (world): do not descend into $(texlinks_dir) if it
	doesn't exist. Report from Andreas Scherer,
	https://tug.org/pipermail/tlbuild/2020q1/004614.html.

2020-05-14  Karl Berry  <karl@freefriends.org>

	* reautoconf for removal of poppler support.

2020-04-10  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2020-03-27  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-27 for TL'20.

2020-03-06  Karl Berry  <karl@tug.org>

	* .travis.yml: attempt to reenable musl build on github.

2020-01-31  Karl Berry  <karl@tug.org>

	* Makefile.am (texlinks_dir): now texk/texlive/tl_support
	(instead of tl_scripts).

2020-01-16  Karl Berry  <karl@freefriends.org>

	* version.ac (tex_live_version): remove /dev, preparing for TL'20.

2019-06-25  Karl Berry  <karl@freefriends.org>

	* Build: don't call buildinfo from here, the output when the CC
	etc. envvars are not set is too misleading. Just write UNAME.

2019-06-24  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh (compiler_version): new shell function;
	 ignore useless first "Configured:" line for gcc->cc on Macs.

2019-06-23  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: output the compiler/other names
	that we pass from the buildinfo target.
	* Makefile.am (buildinfo): pass OBJCXX, not OBJC.
	* Build: more maintainable exports of TL_*.

2019-06-22  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: rename from build-info.sh;
	forget /etc/issue;
	allow option to omit config.status output;
	do not dump environment;
	run lsb_release -a. We'll see ...
	* Build: use buildinfo.sh instead of inline commands (mostly).
	* Makefile.am (buildinfo): make dependency of all-recursive
	so that it runs first; rename from build-info; remove from world.

2019-06-06  Karl Berry  <karl@tug.org>

	* Makefile.am (world): make build-info.
	(build-info): new target to invoke doc/build-info.sh.
	(build_info_target): new variable.
	* Build: export TL_* variables so build-info.sh can report them.
	* doc/build-info.sh: new script to report compiler versions,
	dump environment, config.status, etc.

2019-05-03  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2019-04-07  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): let's hope for 2019-04-07.

2018-10-02  Karl Berry  <karl@tug.org>

	* Build: TL_MAKE_VERBOSE=VERBOSE=1 by default,
	so that test-suite.log will be shown on failure.
	https://lists.gnu.org/archive/html/automake/2018-09/msg00015.html

2018-04-28  Karl Berry  <karl@tug.org>

	* version.ac: 2019/dev.

2018-04-10  Karl Berry  <karl@tug.org>

	* README.*: update from new tlbuild.

2017-11-15  Karl Berry  <karl@tug.org>

	* *.test: use #!/bin/sh -vx. Add 

2017-03-21  Karl Berry  <karl@freefriends.org>

	* everywhere: use EXTRA_DIST=TLpatches instead of <pkgname>-PATCHES.

2017-03-18  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2017/dev => 2017 for TL17.

2016-11-16  Karl Berry  <karl@tug.org>

	* Makefile.am (install-exec-hook): warn that the
	result of make install is not in itself usable.

2016-06-30  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2016 => 2017/dev.
	* tardate.ac (tex_live_tardate): 2015-05-12 => 2016-05-23.
	(Though this is not used for any release.)

2016-03-16  Karl Berry  <karl@tug.org>

	* Build: check for exitstatus.txt not having an exit status.

2016-03-06  Karl Berry  <karl@tug.org>

	* version.ac: remove /dev to prepare for 2016.

2015-06-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2015 => 2016/dev.
	* tardate.ac: Switch to 2015-06-12.

2015-05-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris (removed): Moved from here to doc/.
	* Makefile.am: Adapted.

2015-04-14  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris: New file from Mojca (without very long lines).
	* Makefile.am: Add the new file.

2015-04-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2015-03-07  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2015-03-07.

2014-11-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-11-24.

2014-10-21  Karl Berry  <karl@tug.org>

	* configure.ac: report CC/CXX values, with --version output
	if available.
	* Build: no need for objc any more.

2014-08-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-08-22.

2014-06-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2014 => 2015/dev.
	* tardate.ac: Switch to 2014-06-16.

2014-04-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Makefile.am: New names for README.* files.

2014-04-17  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2014-04-04  Karl Berry  <karl@tug.org>

	* Build (TL_MAKE,MAKE): set both in all cases.
	(buildenv.log): include pwd, and use .log instead of .txt.

2014-04-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-04-01.

2014-01-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-01-01.

2013-12-26  Peter Breitenlohner  <peb@mppmu.mpg.de>


texk/ps2pk/pk2bm.c
/*
 * NAME
 *      pk2bm - create a bitmap from a TeX pkfont
 * SYNOPSIS:
 *	pk2bm [-bh] [-W width] [-H height] {-c char|-o octchar} pkfont 
 * DESCRIPTION
 *	This program generates a bitmap (ASCII file) which can be used
 *	to create X11 applications. The bitmap file consists of all
 *	pixels of the specified character (via the -c or -o option) from 
 *	the given pkfont. The format is described in bitmap(X11).
 *
 *	The pkfont is a packed fontfile generated by gftopk(TeX) from a 
 *	gffont. A gffont is the output of METAFONT a program to design fonts
 *	in a device independant way.
 *
 *	With the -b flag a bitmap is generated in which all black pixels are
 *	drawn as a `*' and all white bits as a `.'. With the -h flag a 
 *	hexadecimal bitmap dump is generated. 
 *
 *      The -W and/or -H options can be used to create a bitmap of
 *	respectivally `width' and `height' pixels. The pk-bitmap will in
 *	this case be centered according to these new dimensions.
 *
 *	The output is written to the standard output.
 * SEE ALSO
 *	``METAFONT'', Donald Knuth.
 *	``The PKtype processor'', belonging to the METAFONTware.
 *	bitmap(X11), gftopk(TeX), pktype(TeX), ps2pk(1)
 * AUTHOR
 *	Piet Tutelaers
 */

#ifdef HAVE_CONFIG_H
#include "c-auto.h"
#endif

#include <stdio.h>
#include "pkin.h"
#include "basics.h"

#ifdef WIN32
#include <fcntl.h>
#include <io.h>
#endif

static quarterword lsbf(quarterword u);
static void dots(quarterword u, int n);
static int atoo(char *oct);

int main(int argc, char *argv[])
{
   int done, C = 0, c, h = 0, w = 0, row, col;
   unsigned int bitmap = 0, hexmap = 0;
   halfword *word;
   chardesc cd;
   const char *myname = "pk2bm";
   char *pkname;
	
   while (--argc > 0 && (*++argv)[0] == '-') {
      done=0;
      while ((!done) && (c = *++argv[0]))	/* allow -bcK as an option */
         switch (c) {
      	 case 'c':
      	    if (*++argv[0] == '\0') {
      	       argc--;  argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
      	    C = *argv[0];
      	    done = 1; break;
      	 case 'o':
      	    if (*++argv[0] == '\0') {
      	       argc--;  ++argv;
      	    }
      	    if (argv[0] == '\0') goto invalid;
      	    C = atoo(argv[0]);
      	    done = 1; break;
      	 case 'H':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
      	    h = atoi(argv[0]);
            done=1;
      	    break;
      	 case 'W':
      	    if (*++argv[0] == '\0') {
      	       argc--; argv++;
      	    }
      	    if (argv[0] == '\0') goto invalid;
      	    w = atoi(argv[0]);
	    done=1;
      	    break;
      	 case 'b':
      	    bitmap = 1;
      	    break;
      	 case 'h':
      	    hexmap = 1;
      	    break;
invalid:
      	 default:
      	    printf("%s: %c invalid option\n", myname, c);
      	    exit(1);
      	 }
   }

   if (argc == 0) {
       msg  ("pk2bm (ps2pk) version " PACKAGE_VERSION "\n");
      printf("Usage: %s [-bh] {-c char|-o octchar} [-W width -H height] pkfile\n", myname);
      exit(1);
   }

#ifdef WIN32
   _setmode(fileno(stdout), _O_BINARY);
#endif

   pkname = argv[0];

   if (readchar(pkname, (integer)C, &cd)) {
      int H, dh, W, dw, bitsleft;
      halfword nextword; quarterword nextbyte;

      H=cd.cheight; if (h==0) h=H; dh=(h-H)/2;
      W=cd.cwidth; if (w==0) w=W; dw=(w-W)/2;
	   
      if (bitmap || hexmap) {
         printf("character : %d (%c)\n", (int) cd.charcode, 
	         (char)cd.charcode);
         printf("   height : %d\n", (int) cd.cheight);
         printf("    width : %d\n", (int) cd.cwidth);
         printf("     xoff : %d\n", (int) cd.xoff);
         printf("     yoff : %d\n", (int) cd.yoff);
      }
      else {
         printf("#define %s_%c_width \t %d\n",
              pkname, (char)cd.charcode, (int) w);
         printf("#define %s_%c_height \t %d\n",
              pkname, (char)cd.charcode, (int) h);
         printf("#define %s_%c_xoff \t %d\n",
              pkname, (char)cd.charcode, (int) dw);
         printf("#define %s_%c_yoff \t %d\n",
              pkname, (char)cd.charcode, (int) dh);
         printf("static char %s_%c_bits[] = {", 
   	      pkname, (char)cd.charcode);
      }
      for (row=0; row<h-H-dh; row++) {
          printf("\n  ");
          for (col=0; col<w; col+=8)
             if (bitmap) dots(0, w-col>7? 8: w-col);
             else if (hexmap) printf("%02x", 0x00);
             else printf(" 0x%02x, ", 0x00);
      }
      word = cd.raster;
      for (row=h-H-dh; row<h-dh; row++) {
         bitsleft = dw; nextword = 0;
         printf("\n  ");
         for (col=0; col<w; col+=8) {
            if (bitsleft>=8) {
               nextbyte = nextword >> (bitsleft-8) & 0xff;
	       bitsleft-=8;
	       
               if (bitmap) dots(nextbyte, w-col>7? 8: w-col); 
               else if (hexmap) printf("%02x", nextbyte);
               else printf(" 0x%02x, ", lsbf(nextbyte));
	    }
	    else {
	       nextbyte = nextword << (8-bitsleft) & 0xff;
	       if (col-dw<W)
	          nextword=(*word++);
	       else nextword=0;
	       nextbyte = nextbyte | (nextword >> (16-(8-bitsleft))
		                           & 0xff);
	       bitsleft = 16 - (8-bitsleft);
	       
               if (bitmap) dots(nextbyte, w-col>7? 8: w-col);
               else if (hexmap) printf("%02x", nextbyte);
               else printf(" 0x%02x, ", lsbf(nextbyte));
	    }
	 }
      }
      for (row=h-dh; row<h; row++) {
          printf("\n  ");
	  for (col=0; col<w; col+=8)
             if (bitmap) dots(0, w-col>7? 8: w-col);
             else if (hexmap) printf("%02x", 0x00);
             else printf(" 0x%02x, ", 0x00);
      }
      if ((!bitmap) && (!hexmap)) printf("};\n");
      else putchar('\n');
   }
   return 0;
}

/*
 * lsbf() transforms a byte (least significant bit first).
 */
static quarterword
lsbf(quarterword u)
{
	int i; quarterword bit, result = 0;

	for (i=0; i<8; i++) {
	   bit = u & 001;
	   result = (result<<1) | bit;
	   u = u >> 1;
	}
	return result;
}

/*
 * dots() print byte: `*' for a black and `.' for a white pixel
 */
static void
dots(quarterword u, int n)
{
   unsigned int bit;

   bit=1<<7;
   for ( ;n>0 ;n--) {
      if (u & bit) putchar('*');
      else putchar('.');
      bit>>=1;
   }
}

static int
atoo(char *oct)
{
   int octal = 0;
   while (*oct != '\0') octal = 8*octal + (*oct++) - '0';
   return octal;
}
